### DISCUSSION TITLE
Different bar margin for max layout

### DISCUSSION THREAD

**Discussion by UberVeng** [qtile/qtile#2843] [Q&A] [answered]
### I want to make different margin for max layout, but it seems like the bar should be updated somehow to apply new gap size.
I wanted to make even gaps, so I made this margin:
```
gapSize = 15

layout_theme = {
    "margin": [0, 0, gapSize, gapSize],
}

bar_config = {
    "margin": gapSize
}
```
And added `bar.Gap` to the right
```
right=bar.Gap(gapSize)
```
I was trying to do something like this: 
```
gapToggle = [0, 15]
gap = cycle(gapToggle)

@lazy.function
def Maximize(qtile):
    gapSize = new(gap),
    qtile.current_group.use_layout(1), // I Don't know how to execute `lazy.next_layout()` here

Key([mod], "c", (Maximize))
```
And it didn't worked. It seems like the variable is changed, but the margins are not updating

#### This is what it looks like in column layout:
![col](https://user-images.githubusercontent.com/23616223/135745954-1ac60782-753f-46dd-854d-4138adf9f3d7.png)

#### And this is in max:
![max](https://user-images.githubusercontent.com/23616223/135745956-abb0fe50-8110-419b-a145-d81d6549b22c.png)

#### And this is how I want it to look like:
![max_](https://user-images.githubusercontent.com/23616223/135745957-1755747e-0ee7-4166-ae58-dcb1a74bc4b4.png)

[config.py.txt](https://github.com/qtile/qtile/files/7273026/config.py.txt)

**Comment by elParaguayo** [comment]
Interesting. I've never tried varying `Bar` margins depending on the layout but I can see why you'd want to do this. 

You'd need a hook to make the adjustment based on the layout.

I might take a look at this as I've played a bit with `Bar` margins when looking at drawing borders on the bar.

**Reply by elParaguayo** [reply]
Actually, it may not be that hard. If you change the bar margins you then just need to reconfigure the bar to recalculate its size and position.

I'm out at the moment but could probably knock out some code for this later today for you to test.

**Comment by elParaguayo** [comment, marked as answer]
Try this as a "proof of concept" (i.e. it can probably be improved...)

``` python
from libqtile import hook, qtile

@hook.subscribe.layout_change
def change_layout(layout, group):

    # Set these as you see fit.
    if layout.name == "max":
        margin = [0, 0, 0, 0]
    else:
        margin = [5, 5, 5, 5]

    try:
        bar = qtile.current_screen.top

    # I think current_screen may not be defines when the hook is first fired so this will just catch any error
    except AttributeError:
        return

    bar.margin = margin
    bar._configure(qtile, qtile.current_screen)
    bar.draw()
```

**Reply by UberVeng** [reply]
That works perfectly fine! Thank you! One last thing: the bar on the right is still there. I can see that you used `bar.margin` to change margin,  so is there a way to change the size of the right `bar.Gap`?

**Reply by elParaguayo** [reply]
OK. Try this:

``` python
from libqtile import hook, qtile
from libqtile.bar import Gap

@hook.subscribe.layout_change
def change_layout(layout, group):

    # Set these as you see fit.
    if layout.name == "max":
        margin = [0, 0, 0, 0]
        gap = None
    else:
        margin = [5, 5, 5, 5]
        gap = Gap(5)

    try:
        bar = qtile.current_screen.top

    # I think current_screen may not be defines when the hook is first fired so this will just catch any error
    except AttributeError:
        return

    qtile.current_screen.right = gap
    bar.margin = margin
    bar._configure(qtile, qtile.current_screen)
    bar.draw()
    qtile.current_screen.group.layout_all()
```

Haven't tested but should be something like that.

**Reply by UberVeng** [reply]
I really appreciate your help @elParaguayo! I was trying to find the solution for so long, that I started to thinking that it's not possible it qtile. Thank you again!

**Reply by gruvw** [reply]
Hello, I can't understand why but this is not working for me...

**Reply by elParaguayo** [reply]
I don't know either!

You might need to share some more details about what you've done (including your config) and what's not working.

**Reply by gruvw** [reply]
Here is my config https://github.com/gruvw/dotfiles/blob/main/qtile/.config/qtile/config.py
When I switch to the max layout, the top-bar margins do not change.

**Reply by elParaguayo** [reply]
Change these two lines:
``` python
    bar.margin = margin
    bar._configure(qtile, qtile.current_screen)
```
to
``` python
    bar._initial_margin = margin
    bar._configure(qtile, qtile.current_screen, reconfigure=True)
```

**Reply by gruvw** [reply]
I made the change but it still does not work...

**Reply by elParaguayo** [reply]
Are you sure? Works perfectly for me. Can you post screenshot of Max and another layout?

**Reply by elParaguayo** [reply]
Did you restart qtile?

**Reply by gruvw** [reply]
Ok I got it.
I was not using the max layout in fact, I was using `lazy.window.toggle_maximize()` thinking that this was using the max layout but it was not.
Do you know how to have a shortcut trigger the use of the max layout ?

**Reply by elParaguayo** [reply]
`lazy.group.setlayout("max")`

**Reply by gruvw** [reply]
great thank you so much for your help!
